hamilton/lifecycle/caching.py

```diff
@@ -0,0 +1,1087 @@
+        """The data_version of override values must be computed to enable cache checks.
+
+        Don't store overrides in metadata_store because their value isn't actually tied to the code.
+        There's little reason to store the value in the result_store since their would be no keys
+        in the metadata_store to retrieve it.
+        """
+        data_version = self.version_data(node_name=node_name, result=value)
+        self._log_event(
+            node_name=node_name,
+            task_id=None,
+            actor="adapter",
+            event_type=CachingEventType.AS_OVERRIDE,
+            msg="",
+            value=data_version,
+        )
+        self.set_memory_metadata(node_name=node_name, data_version=data_version)
+
+    def run_before_graph_execution(
+        self,
+        *,
+        run_id: str,
+        graph: graph_types.HamiltonGraph,
+        inputs: Dict[str, Any],
+        overrides: Dict[str, Any],
+        execution_path: Collection[str],
+        **kwargs: Any,
+    ):
+        """Get code versions for all, and data version for top-level inputs and overrides
+        and setup state variables before the run
+        """
+        self.run_id = run_id
+        self._logs[run_id] = []
+        self.metadata_store.initialize()
+
+        self.graph = graph
+        self.resolve_behaviors(graph=graph)
+        # reset the code_versions and data versions before each execution
+        self.data_versions = {}
+        # code versions aren't expected to change for the lifetime of the Driver though
+        self.code_versions = {
+            node.name: self.version_code(node_name=node.name, node=node)
+            for node in self.graph.nodes
+            if node.name in execution_path
+        }
+
+        if inputs:
+            for node_name, value in inputs.items():
+                self._process_input(node_name, value)
+
+        if overrides:
+            for node_name, value in overrides.items():
+                self._process_override(node_name, value)
+
+    def run_before_node_execution(self, *args, **kwargs):
+        """required to subclass ``NodeExecutionHook``"""
+
+    # subclass `BaseDoNodeExecute` instead of `NodeExecutionMethod` to access `node.Node` object
+    def do_node_execute(
+        self,
+        *,
+        node_: hamilton.node.Node,
+        kwargs: Dict[str, Any],
+        task_id: Optional[str] = None,
+        **future_kwargs,
+    ):
+        """If the metadata store includes the context key, else compute result.
+        If a fingerprint is retrieved, but reading from cache fails, invalidate the metadata store
+        and force recompute and storage of fingerprint + result.
+        """
+        # gather the attributes typically available through NodeExecutionMethod
+        node_name = node_.name
+        node_callable = node_.callable
+        node_kwargs = kwargs
+
+        if node_.node_role == hamilton.node.NodeType.EXPAND:
+            self._expand_nodes.add(node_name)
```

Comment:
Why can't we do this at the beginning? We know all the node types statically, pre-graph-execute could work.

Reply:
I tried looking at `HamiltonNode` but I'm pretty sure they don't communicate `Expand` / `Collect`. I could manually investigate the type again though

Reply:
You can use the function graph directly: https://github.com/DAGWorks-Inc/hamilton/blob/72866175bb1e5dc65cedfd2de1096cc082fa19ee/hamilton/lifecycle/api.py#L300